iothub/device/src/InternalClient.cs

```diff
@@ -121,7 +122,9 @@ internal class InternalClient : IDisposable
         private volatile Tuple<MessageHandler, object> _defaultEventCallback;
 
         // Stores methods supported by the client device and their associated delegate.
-        private volatile Dictionary<string, Tuple<MethodCallback, object>> _deviceMethods;
+
+        private volatile bool _isDeviceMethodEnabled;
+        private readonly Dictionary<string, Tuple<MethodCallback, object>> _deviceMethods = new Dictionary<string, Tuple<MethodCallback, object>>();
```

Comment:
I'm not very sure if this should be volatile or not. But just an observation that only the bool is volatile now. I have never used volatile in my coding experience but if you have an answer, I would like to learn why it was only used on the bool.

Reply:
It means it could be changed by more than one thread, and the compiler will remove optimizations that might otherwise compromise reading a fresh value.

<https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/volatile>

Reply:
Checking again, this boolean is only changed inside a semaphore, so I think we don't need it.